grpc/src/main/java/io/stargate/grpc/service/Service.java

```diff
@@ -204,42 +204,35 @@ private void executePrepared(
       Connection connection,
       Prepared prepared,
       Query query,
-      StreamObserver<Result> responseObserver) {
+      StreamObserver<Response> responseObserver) {
     try {
       long queryStartNanoTime = System.nanoTime();
 
-      QueryParameters parameters = query.getParameters();
-      Payload payload = parameters.getPayload();
+      Payload values = query.getValues();
+      PayloadHandler handler = PayloadHandlers.get(values.getType());
 
-      PayloadHandler handler = PayloadHandlers.get(payload.getType());
+      QueryParameters parameters = query.getParameters();
 
       connection
           .execute(
-              bindValues(handler, prepared, payload),
-              makeParameters(parameters),
-              queryStartNanoTime)
+              bindValues(handler, prepared, values), makeParameters(parameters), queryStartNanoTime)
           .whenComplete(
               (result, t) -> {
                 if (t != null) {
                   handleError(t, responseObserver);
                 } else {
                   try {
-                    Result.Builder resultBuilder = makeResultBuilder(result);
+                    Response.Builder responseBuilder = makeResponseBuilder(result);
                     switch (result.kind) {
                       case Void:
                       case SchemaChange: // Fallthrough intended
                         break;
                       case Rows:
-                        Rows rows = (Rows) result;
-                        if (rows.rows.isEmpty()
-                            && (parameters.getSkipMetadata()
-                                || rows.resultMetadata.columns.isEmpty())) {
-                          resultBuilder.setPayload(
-                              Payload.newBuilder().setType(payload.getType()).build());
-                        } else {
-                          resultBuilder.setPayload(
-                              handler.processResult((Rows) result, parameters));
-                        }
+                        responseBuilder.setResultSet(
```

Comment:
This now just returns an empty payload data for empty result sets. Which is both less complex and more consistent.